contracts/governance/extensions/GovernorSuperQuorum.sol

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Governor} from "../Governor.sol";
+import {IGovernorCounting} from "./IGovernorCounting.sol";
+import {SafeCast} from "../../utils/math/SafeCast.sol";
+import {Checkpoints} from "../../utils/structs/Checkpoints.sol";
+
+/**
+ * @dev Extension of {Governor} with a super quorum. Proposals that meet the super quorum can be executed
+ * earlier than the proposal deadline. Counting modules that want to use this extension must implement
+ * {IGovernorCounting}.
+ *
+ * NOTE: It's up to developers to implement `superQuorum` and validate it against `quorum`.
+ */
+abstract contract GovernorSuperQuorum is Governor {
+    /**
+     * @dev Minimum number of cast votes required for a proposal to reach super quorum. Only FOR votes are counted towards
+     * the super quorum. Once the super quorum is reached, an active proposal can proceed to the next state without waiting
+     * for the proposal deadline.
+     *
+     * NOTE: The `timepoint` parameter corresponds to the snapshot used for counting vote. This allows to scale the
+     * quorum depending on values such as the totalSupply of a token at this timepoint (see {ERC20Votes}).
+     */
+    function superQuorum(uint256 timepoint) public view virtual returns (uint256);
+
+    /**
+     * @dev Overridden version of the {Governor-state} function that checks if the proposal has reached the super quorum.
+     *
+     * NOTE: If the proposal reaches super quorum but {_voteSucceeded} returns false, eg, assuming the super quorum has been set
+     * low enough that both FOR and AGAINST votes have exceeded it and AGAINST votes exceed FOR votes, the proposal continues to
+     * be active until {_voteSucceeded} returns true or the proposal deadline is reached. This means that with a low super quorum
+     * it is also possible that a vote can succeed prematurely before enough AGAINST voters have a chance to vote. Hence, it is
+     * recommended to set a high enough super quorum to avoid these types of scenarios.
+     */
+    function state(uint256 proposalId) public view virtual override returns (ProposalState) {
+        ProposalState currentState = super.state(proposalId);
+        if (currentState != ProposalState.Active) return currentState;
+
+        (, uint256 forVotes, ) = IGovernorCounting(address(this)).proposalVotes(proposalId);
```

Comment:
I understand this function is external in the interface, but the intuition of calling `this` with `proposalVotes` ignores that `proposalVotes` is public in every Governor counting module. 

Given the compiler doesn't give an error when you mark as `public` a function that is `external` in the interface, I think that should be the default since it's the least disruptive action. Otherwise I'm worried that it replaces the execution environment (i.e. msg.sender and others), which could be relevant.

I'd use the `IGovernorCounting` interface and:

```suggestion
        (, uint256 forVotes, ) = proposalVotes(proposalId);
```

Reply:
> GovernorSuperQuorum has no knowledge of the counting modules/proposalVotes as that's not inherited from Governor so I cannot drop the casting from here.

Its easy to give him knowledge of that by defining the (unimplemented) function as public here.